### ISSUE TITLE
cosmopolitan libc for "αcτµαlly pδrταblε εxεcµταblε"s

### ISSUE THREAD

**Issue by ihsanturk** [NixOS/nixpkgs#113195]
**Project description**
> Cosmopolitan makes C a build-once run-anywhere language, similar to Java, except it doesn't require interpreters or virtual machines be installed beforehand. Cosmo provides the same portability benefits as high-level languages like Go and Rust, but it doesn't invent a new language and you won't need to configure a CI system to build separate binaries for each operating system. What Cosmopolitan focuses on is fixing C by decoupling it from platforms, so it can be pleasant to use for writing small unix programs that are easily distributed to a much broader audience.

**Metadata**

* homepage URL: https://justine.lol/cosmopolitan/index.html
* source URL: https://github.com/jart/cosmopolitan
* license: ISC license
* platforms: Unix, Linux, Darwin, Windows

**Comment by lourkeur** [comment]
https://github.com/jart/cosmopolitan/issues/63

**Comment by lourkeur** [comment]
I have been making some progress on that on [my branch](https://github.com/lourkeur/nixpkgs/blob/cosmopolitan/pkgs/development/compilers/cosmopolitan/default.nix). I might be able to PR it once 0.3 drops.

**Comment by tomberek** [comment]
MIght be good to have this as a variant of pkgsStatic.

**Comment by lourkeur** [comment]
> MIght be good to have this as a variant of pkgsStatic.

There's a potential issue in that the αpε format bootstrap itself by editing the first kB of the executable, which we cant do in the store. We might have to write a different bootstrap mechanism, but we can still use the libc and all the cross-platform work. That's not to say we shouldn't offer stock cosmo alongside that, that's awesome too!

**Comment by ihsanturk** [comment]
Yes I noticed that too, when I compile the program and run the binary once, it changes itself to only be compatible with that operating system and the binary is not portable anymore. It can only be portable when it hasn't been executed yet.

**Comment by tomberek** [comment]
A hack around the problem:

```
#!@shell@
cp @original@ $TMP/@name@
chmod +w $TMP/@name@
exec @name@ "$@"
```

**Comment by ihsanturk** [comment]
Looks cool

**Comment by pniedzwiedzinski** [comment]
> A hack around the problem:
> 
> ```
> #!@shell@
> cp @original@ $TMP/@name@
> chmod +w $TMP/@name@
> exec @name@ "$@"
> ```

This makes it less reproducible though. Another idea is to just run the executable during the install phase, but I'm not sure how to safely do this. 

NixOS packages are not meant to be portable, as the same package on different systems will still produce different nar/hash. Still, having ability to produce cross platform programs without setting up pipelines is still a great idea and I think this is what we can leverage from cosmopolitan.

**Comment by OPNA2608** [comment]
Package was merged into master via #114628, seems done to me.
